projects/ad469x_fmcz/src/ad469x_fmcz.c

```diff
@@ -0,0 +1,190 @@
+#include "pwm.h"
+#include "axi_pwm_extra.h"
+#include "ad469x.h"
+#include "error.h"
+#include "delay.h"
+#include "clk_axi_clkgen.h"
+#include "gpio.h"
+#include "gpio_extra.h"
+
+/******************************************************************************/
+/********************** Macros and Constants Definitions **********************/
+/******************************************************************************/
+#define AD469x_EVB_SAMPLE_NO		1000
+#define AD469x_DMA_BASEADDR             XPAR_AXI_AD4696_DMA_BASEADDR
+#define AD469x_SPI_ENGINE_BASEADDR      XPAR_SPI_AD4696_AXI_REGMAP_BASEADDR
+#define RX_CLKGEN_BASEADDR		XPAR_SPI_CLKGEN_BASEADDR
+#define AXI_PWMGEN_BASEADDR		XPAR_SPI_AD4696_TRIGGER_GEN_BASEADDR
+#define AD469x_SPI_ENG_REF_CLK_FREQ_HZ	XPAR_PS7_SPI_0_SPI_CLK_FREQ_HZ
+#define AD469x_SPI_CS                   0
+#define GPIO_OFFSET			54
+#define GPIO_RESETN_1			GPIO_OFFSET + 32
+#define GPIO_DEVICE_ID			XPAR_PS7_GPIO_0_DEVICE_ID
+
+int main()
+{
+	struct ad469x_dev *dev;
+	uint32_t *offload_data;
+	struct spi_engine_offload_message msg;
+	uint32_t commands_data[1];
+	int32_t ret, data;
+	uint32_t i, j = 0;
+
+	struct spi_engine_offload_init_param spi_engine_offload_init_param = {
+		.offload_config = OFFLOAD_RX_EN,
+		.rx_dma_baseaddr = AD469x_DMA_BASEADDR,
+	};
+
+	struct spi_engine_init_param spi_eng_init_param  = {
+		.ref_clk_hz = AD469x_SPI_ENG_REF_CLK_FREQ_HZ,
+		.type = SPI_ENGINE,
+		.spi_engine_baseaddr = AD469x_SPI_ENGINE_BASEADDR,
+		.cs_delay = 0,
+		.data_width = 16,
+	};
+
+	struct axi_clkgen_init clkgen_init = {
+		.name = "rx_clkgen",
+		.base = RX_CLKGEN_BASEADDR,
+		.parent_rate = 100000000,
+	};
+
+	struct axi_pwm_init_param axi_pwm_init = {
+		.base_addr = AXI_PWMGEN_BASEADDR,
+		.ref_clock_Hz = 160000000,
+	};
+
+	struct pwm_init_param trigger_pwm_init = {
+		.period_ns = 1000,	/* 1Mhz */
+		.duty_cycle_ns = 10,
+		.polarity = PWM_POLARITY_HIGH,
+		.extra = &axi_pwm_init,
+	};
+
+	struct xil_gpio_init_param gpio_extra_param = {
+		.device_id = GPIO_DEVICE_ID,
+		.type = GPIO_PS,
+	};
+
+	struct gpio_init_param ad469x_resetn = {
+		.number = GPIO_RESETN_1,
+		.extra = &gpio_extra_param
+	};
+
+	struct spi_init_param spi_init = {
+		.chip_select = AD469x_SPI_CS,
+		.max_speed_hz = 80000000,
+		.mode = SPI_MODE_3,
+		.platform_ops = &spi_eng_platform_ops,
+		.extra = (void*)&spi_eng_init_param,
+	};
+
+	struct ad469x_init_param ad469x_init_param = {
+		.spi_init = spi_init,
+		.clkgen_init = clkgen_init,
+		.axi_clkgen_rate = 160000000,
+		.trigger_pwm_init = trigger_pwm_init,
+		.reg_access_speed = 20000000,
+		.reg_data_width = 8,
+		.capture_data_width = 16,
+		.dev_id = ID_AD4696, /* dev_id */
+		.gpio_resetn = &ad469x_resetn,
+	};
+
+	uint32_t spi_eng_msg_cmds[3] = {
+		CS_LOW,
+		WRITE_READ(1),
+		CS_HIGH
+	};
+
+	Xil_ICacheEnable();
+	Xil_DCacheEnable();
+
+	ret = ad469x_init(&dev, &ad469x_init_param);
+	if (ret < 0)
+		return ret;
+
+	ret = spi_engine_offload_init(dev->spi_desc,
```

Comment:
Should we access spi_desc from dev in main? 
Isn't it better to have a ad469x_read and ad469x_write functions?